sqlite.go

```diff
@@ -0,0 +1,542 @@
+// This is useful but generates extra string allocations.
+// Turn it off if you have written a program with an SQL error on a hot path.
+var ErrMsg = true
+
+var _ driver.Driver = Driver{}
+var _ driver.DriverContext = Driver{}
+var _ driver.ConnBeginTx = (*conn)(nil)
+var _ driver.StmtExecContext = (*stmt)(nil)
+var _ driver.StmtQueryContext = (*stmt)(nil)
+
+var drv Driver
+
+type Driver struct{}
+
+func (d Driver) Open(name string) (driver.Conn, error) { panic("unused") }
+func (d Driver) OpenConnector(name string) (driver.Connector, error) {
+	return &connector{name: name}, nil
+}
+
+type connector struct {
+	name string
+}
+
+func (c *connector) Driver() driver.Driver { return drv }
+func (c *connector) Connect(ctx context.Context) (driver.Conn, error) {
+	log.Printf("connector.Connect: %s", c.name)
+	flags := sqliteh.SQLITE_OPEN_URI |
+		sqliteh.SQLITE_OPEN_READWRITE |
+		sqliteh.SQLITE_OPEN_CREATE
+	db, err := cgosqlite.Open(c.name, flags, "")
+	if err != nil {
+		e := &Error{
+			Code: sqliteh.Code(err.(sqliteh.ErrCode)),
+			Loc:  "Open",
+		}
+		if db != nil {
+			e.Msg = db.ErrMsg()
+			db.Close()
+		}
+		log.Printf("connector.Connect: err=%v", err)
+		return nil, e
+	}
+	return &conn{db: db}, nil
+}
+
+type conn struct {
+	db    *cgosqlite.DB
+	stmts map[string]*stmt // persisted statements
+}
+
+func (c *conn) Prepare(query string) (driver.Stmt, error) { panic("unused") }
+func (c *conn) Begin() (driver.Tx, error)                 { panic("unused") }
+func (c *conn) Close() error {
+	return reserr(c.db, "Conn.Close", "", c.db.Close())
+}
+func (c *conn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	return c.prepare(query, false)
+}
+
+func (c *conn) prepare(query string, persist bool) (*stmt, error) {
+	if s := c.stmts[query]; s != nil {
+		return s, nil
+	}
+	cstmt, rem, err := c.db.Prepare(query, 0)
+	if err != nil {
+		return nil, reserr(c.db, "Prepare", query, err)
+	}
+	if rem != "" {
+		cstmt.Finalize()
+		return nil, &Error{
+			Code:  sqliteh.SQLITE_MISUSE,
+			Loc:   "Prepare",
+			Query: query,
+			Msg:   fmt.Sprintf("query has trailing text: %q", rem),
+		}
+	}
+	s := &stmt{db: c.db, stmt: cstmt, query: query, persist: persist}
+	if persist {
+		if c.stmts == nil {
+			c.stmts = make(map[string]*stmt)
+		}
+		c.stmts[query] = s
+	}
+	return s, nil
+}
+
+func (c *conn) execInternal(ctx context.Context, query string) error {
+	s, err := c.prepare(query, true)
+	if err != nil {
+		if e, _ := err.(*Error); e != nil {
+			e.Loc = "internal:" + e.Loc
+		}
+		return err
+	}
+	if _, err := s.ExecContext(ctx, nil); err != nil {
+		return err
+	}
+	s.Close()
+	return nil
+}
+
+func (c *conn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx, error) {
+	const LevelSerializable = 6 // matches the sql package constant
+	if opts.Isolation != 0 && opts.Isolation != LevelSerializable {
+		return nil, errors.New("sqlite driver only supports serializable isolation level")
```

Comment:
Does it not support linearizable as well? AFAICT sqlite's serializability translates to linearizability as well, but I'm bad at grokking edge cases.

Reply:
I think sqlite is linearizable, but I haven't proven it to myself. The sqlite docs explicitly call out serializable, so that's what I went with: https://sqlite.org/isolation.html